# Data Cleaning

In [1]:
import folium
import pandas as pd
import seaborn as sns
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf
import seaborn as sns
import numpy as np

In [2]:
crash = pd.read_csv("Motor_Vehicle_Collisions_-_Crashes.csv", na_values = ["Unspecified"])
pd.set_option('display.max_columns', None)
crash.head()

/usr/local/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


CRASH DATE CRASH TIME BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0  03/06/2020       6:56     NaN      NaN  40.687004 -73.917690   
1  04/28/2020      17:35  QUEENS    11692  40.589134 -73.800980   
2  03/18/2020      14:00     NaN      NaN  40.666260 -73.995960   
3  05/07/2020      22:14     NaN      NaN  40.787224 -73.954170   
4  03/06/2020      17:16     NaN      NaN  40.702007 -73.821205   

                  LOCATION                    ON STREET NAME  \
0   (40.687004, -73.91769)  BROADWAY                           
1   (40.589134, -73.80098)  ROCKAWAY BEACH BOULEVARD           
2    (40.66626, -73.99596)  GOWANUS EXPY (BQE)                 
3   (40.787224, -73.95417)  MADISON AVENUE                     
4  (40.702007, -73.821205)  131 STREET                         

  CROSS STREET NAME OFF STREET NAME  NUMBER OF PERSONS INJURED  \
0  JEFFERSON AVENUE             NaN                        0.0   
1   BEACH 73 STREET             NaN                        0.0   
2               NaN             NaN                        1.0   
3               NaN             NaN                        0.0   
4    JAMAICA AVENUE             NaN                        0.0   

   NUMBER OF PERSONS KILLED  NUMBER OF PEDESTRIANS INJURED  \
0                       0.0                              0   
1                       0.0                              0   
2                       0.0                              0   
3                       0.0                              0   
4                       0.0                              0   

   NUMBER OF PEDESTRIANS KILLED  NUMBER OF CYCLIST INJURED  \
0                             0                          0   
1                             0                          0   
2                             0                          0   
3                             0                          0   
4                             0                          0   

   NUMBER OF CYCLIST KILLED  NUMBER OF MOTORIST INJURED  \
0                         0                           0   
1                         0                           0   
2                         0                           1   
3                         0                           0   
4                         0                           0   

   NUMBER OF MOTORIST KILLED   CONTRIBUTING FACTOR VEHICLE 1  \
0                          0             Alcohol Involvement   
1                          0  Driver Inattention/Distraction   
2                          0                Brakes Defective   
3                          0  Driver Inattention/Distraction   
4                          0   Failure to Yield Right-of-Way   

    CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3  \
0                             NaN                           NaN   
1  Driver Inattention/Distraction                           NaN   
2                             NaN                           NaN   
3                             NaN                           NaN   
4                             NaN                           NaN   

  CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
0                           NaN                           NaN       4305304   
1                           NaN                           NaN       4310784   
2                           NaN                           NaN       4302030   
3                           NaN                           NaN       4312375   
4                           NaN                           NaN       4305442   

                   VEHICLE TYPE CODE 1                  VEHICLE TYPE CODE 2  \
0  Station Wagon/Sport Utility Vehicle                                  NaN   
1  Station Wagon/Sport Utility Vehicle  Station Wagon/Sport Utility Vehicle   
2  Station Wagon/Sport Utility Vehicle                                Sedan   
3  Station Wagon/Sport Utility Vehicle                            Ambulance   
4  Station Wagon/Sport Utility Vehicle                     

In [3]:
new_columns = ["date","time","borough","zipcode", "latitude", "longitude", "location", "on_street", "cross_street", "off_street", "num_persons_injured", "num_persons_killed", "num_ped_injured", "num_ped_killed", "num_cyc_injured", "num_cyc_killed", "num_mot_injured", "num_mot_killed", "con_vehicle_1", "con_vehicle_2", "con_vehicle_3", "con_vehicle_4", "con_vehicle_5", "collision_id", "vehicle_type_1", "vehicle_type_2", "vehicle_type_3", "vehicle_type_4", "vehicle_type_5"]
crash.columns = new_columns

In [4]:
crash["date"] = pd.to_datetime(crash["date"])
crash.head()

date   time borough zipcode   latitude  longitude  \
0 2020-03-06   6:56     NaN     NaN  40.687004 -73.917690   
1 2020-04-28  17:35  QUEENS   11692  40.589134 -73.800980   
2 2020-03-18  14:00     NaN     NaN  40.666260 -73.995960   
3 2020-05-07  22:14     NaN     NaN  40.787224 -73.954170   
4 2020-03-06  17:16     NaN     NaN  40.702007 -73.821205   

                  location                         on_street  \
0   (40.687004, -73.91769)  BROADWAY                           
1   (40.589134, -73.80098)  ROCKAWAY BEACH BOULEVARD           
2    (40.66626, -73.99596)  GOWANUS EXPY (BQE)                 
3   (40.787224, -73.95417)  MADISON AVENUE                     
4  (40.702007, -73.821205)  131 STREET                         

       cross_street off_street  num_persons_injured  num_persons_killed  \
0  JEFFERSON AVENUE        NaN                  0.0                 0.0   
1   BEACH 73 STREET        NaN                  0.0                 0.0   
2               NaN        NaN                  1.0                 0.0   
3               NaN        NaN                  0.0                 0.0   
4    JAMAICA AVENUE        NaN                  0.0                 0.0   

   num_ped_injured  num_ped_killed  num_cyc_injured  num_cyc_killed  \
0                0               0                0               0   
1                0               0                0               0   
2                0               0                0               0   
3                0               0                0               0   
4                0               0                0               0   

   num_mot_injured  num_mot_killed                   con_vehicle_1  \
0                0               0             Alcohol Involvement   
1                0               0  Driver Inattention/Distraction   
2                1               0                Brakes Defective   
3                0               0  Driver Inattention/Distraction   
4                0               0   Failure to Yield Right-of-Way   

                    con_vehicle_2 con_vehicle_3 con_vehicle_4 con_vehicle_5  \
0                             NaN           NaN           NaN           NaN   
1  Driver Inattention/Distraction           NaN           NaN           NaN   
2                             NaN           NaN           NaN           NaN   
3                             NaN           NaN           NaN           NaN   
4                             NaN           NaN           NaN           NaN   

   collision_id                       vehicle_type_1  \
0       4305304  Station Wagon/Sport Utility Vehicle   
1       4310784  Station Wagon/Sport Utility Vehicle   
2       4302030  Station Wagon/Sport Utility Vehicle   
3       4312375  Station Wagon/Sport Utility Vehicle   
4       4305442  Station Wagon/Sport Utility Vehicle   

                        vehicle_type_2                       vehicle_type_3  \
0                                  NaN                                  NaN   
1  Station Wagon/Sport Utility Vehicle                                  NaN   
2                                Sedan  Station Wagon/Sport Utility Vehicle   
3                            Ambulance                                  NaN   
4                                  NaN                                  NaN   

  vehicle_type_4 vehicle_type_5  
0            NaN            NaN  
1            NaN            NaN  
2            NaN            NaN  
3            NaN            NaN  
4            NaN            NaN

In [5]:
crash["date"] = crash["date"].dt.year

In [11]:
crash = crash[["date","latitude", "longitude"]]

In [ ]:
crash

In [ ]:
crash["date"].value_counts()

# Maps

In [12]:
filt1 = crash["date"] == 2020
crash2020 = crash[filt1]
crash2020.head()

date   latitude  longitude
0  2020  40.687004 -73.917690
1  2020  40.589134 -73.800980
2  2020  40.666260 -73.995960
3  2020  40.787224 -73.954170
4  2020  40.702007 -73.821205

In [13]:
crash2020.loc[50:55]

date   latitude  longitude
50  2020  40.656685  -73.92990
51  2020  40.848480  -73.92249
53  2020  40.736350  -73.97502
54  2020  40.713135  -74.00407
55  2020  40.713955  -73.95794

In [14]:
crash2020.reset_index(inplace = True, drop = True)

In [15]:
len(crash2020)

112834

In [24]:
sample100 = crash2020.sample(100)

In [25]:
sample100 = sample100.dropna()

In [26]:
sample100.reset_index(inplace = True, drop = True)

In [27]:
crash_map = folium.Map(location=[40.7128, -74.0060])
crash_map

In [28]:
for i in range(len(sample100)):
    row = sample100.loc[i]
    print("Coordinates: ", row["latitude"], row["longitude"] )

Coordinates:  40.702335 -73.89073
Coordinates:  40.702686 -73.92084
Coordinates:  40.6665 -73.9962
Coordinates:  40.825428 -73.93630999999999
Coordinates:  40.83932 -73.91503
Coordinates:  40.643603999999996 -73.94301
Coordinates:  40.8076 -73.93719
Coordinates:  40.549503 -74.21613
Coordinates:  40.65463 -73.86055
Coordinates:  40.683437 -73.92917
Coordinates:  40.675816 -73.895836
Coordinates:  40.657383 -73.911514
Coordinates:  40.817875 -73.89333
Coordinates:  40.67685 -73.981575
Coordinates:  40.751521999999994 -73.99396
Coordinates:  40.76296 -73.887665
Coordinates:  40.778605999999996 -73.98163000000001
Coordinates:  40.822346 -73.879616
Coordinates:  40.844967 -73.89705
Coordinates:  40.575540000000004 -73.97964499999999
Coordinates:  40.687126 -73.99361400000001
Coordinates:  40.667770000000004 -73.94514000000001
Coordinates:  40.669533 -73.91181
Coordinates:  40.826515 -73.917946
Coordinates:  40.70145 -73.98962399999999
Coordinates:  40.643917 -74.0008
Coordinates:  40.67863

In [ ]:
crash_map = folium.Map(location=[40.7128, -74.0060])
for i in range(len(sample100)):
    row = sample100.loc[i]
    folium.Marker([row["latitude"], row["longitude"]], 
                 # tooltip=row["on_street"]
             ).add_to(crash_map)

In [31]:
crash_map

Generating this map took multiple steps. In addition to the steps I completed in the initial cleaning, I filtered the data so that I only have car crashes that happened in 2020. Additionally, I sliced that dataset so that I only had three columns, date, latitude, and longitude. I then took a sample of 100 from the 112,000 collisions that occurred last year because it was impossible to plot all those points on this map. 

The first step I took to create the maps was to create an empty map of New York City. Next, I used iteration, or loops, to loop through all the rows of the dataset, store the rows in the variable "row", and used it to print the latitude and longitude. Lastly, I plotted the crashes on the map. The latitudes and longitudes from each row were used to create all the markers for my map. 

Pictured below/above is a map of New York City. Plotted on the map are 100 of the 112,834 collisions that occurred in 2020. There are 20 in the Bronx, 14 in Manhattan, and 2 in Staten Island. Between Brooklyn and Manhattan, there were 64 crashes (I cannot tell where the boundaries for the two boroughs are). 